river/cluster/clustream.py

```diff
@@ -0,0 +1,363 @@
+                    M=self.max_kernels,
+                )
+
+                return self
+
+        """try to merge closest two kernels"""
+        closest_a = 0
+        closest_b = 0
+        min_distance = sys.float_info.max
+        for i, kernel_i in self.kernels.items():
+            center_a = kernel_i.center
+            for j, kernel_j in self.kernels.items():
+                dist = self._distance(center_a, kernel_j.center)
+                if dist < min_distance and j > i:
+                    min_distance = dist
+                    closest_a = i
+                    closest_b = j
+        self.kernels[closest_a].add(self.kernels[closest_b])
+        self.kernels[closest_b] = ClustreamKernel(
+            x=x,
+            sample_weight=sample_weight,
+            timestamp=self.time_stamp,
+            T=self.kernel_radius_factor,
+            M=self.max_kernels,
+        )
+
+        return self
+
+    def get_micro_clustering_result(self):
+
+        if not self.initialized:
+            return {}
+        res = {
+            i: ClustreamKernel(
+                cluster=self.kernels[i], T=self.kernel_radius_factor, M=self.max_kernels
+            )
+            for i in range(len(self.kernels))
+        }
+        return res
+
+    def learn_predict_one(self, x, sample_weight=None):
+        """
+        Compute cluster centers and predict cluster index for each sample.
+
+        Convenience method; equivalent to calling learn_one(x) followed by predict_one(x).
+
+        Parameters
+        ----------
+        x
+            Instance attributes
+
+        sample_weight
+            Instance weights. If not provided, uniform weights are assumed
+
+        Returns
+        -------
+        y
+            Integer
+            Cluster label
+        """
+
+        self.learn_one(x, sample_weight)
+
+        y = self.predict_one(x)
+
+        return y
+
+    @staticmethod
+    def _get_closest_kernel(x, micro_clusters):
+        """
+
+        Parameters
+        ----------
+        x
+            A dictionary with n items (features).
+            Instance attributes. Due to the limitations of the dictionary, x on default is considered as one instance
+
+        micro_clusters
+            Dictionary like
+            Instance weight. If not provided, uniform weights are assumed
+
+        Returns
+        -------
+        closest_kernel_index : int
+            Index of closest kernel to the given instance
+
+        min_distance: float
+            distance between closest kernel to the given instance
+        """
+
+        min_distance = sys.float_info.max
+        closest_kernel_index = -1
+        for i, micro_cluster in micro_clusters.items():
+            distance = 0
+            for j in range(len(x)):
+                distance += (micro_cluster.center[j] - x[j]) * (
+                    micro_cluster.center[j] - x[j]
+                )
+            distance = math.sqrt(distance)
```

Comment:
Is there any reason why this is not using the `_distance` static method?

Reply:
As I have defined it as a static method, so if we want to use the `_distance` function, we have to pass `self` into the function, which makes the `_get_closest_kernel` not a static method anymore.